plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt

```diff
@@ -0,0 +1,181 @@
+    val fullMethod: Method get() = signature.toMethod(listOf(), listOf(), convertedBody)
+    val headerOnlyMethod: Method get() = signature.toMethod(listOf(), listOf(), null)
+
+    val returnVar: ConvertedVar?
+    val signature: ConvertedMethodSignature
+
+    init {
+        val retType = (declaration.returnTypeRef as FirResolvedTypeRef).type
+        val convertedRetType = programCtx.convertType(retType)
+        returnVar = (if (convertedRetType is ConvertedNonUnitType) ConvertedVar(RETURN_VARIABLE_NAME, convertedRetType) else null)
+
+        val params = declaration.valueParameters.map {
+            ConvertedVar(
+                it.name.toString(),
+                programCtx.convertType((it.returnTypeRef as FirResolvedTypeRef).type) as ConvertedNonUnitType
+            )
+        }
+        val returns = returnVar?.let { listOf(it) } ?: emptyList()
+        signature = ConvertedMethodSignature(declaration.name.asString(), params, returns)
+    }
+
+    private val convertedBody: Seqn
+        get() {
+            val body = declaration.body ?: throw Exception("Functions without a body are not supported yet.")
+            val ctx = StmtConversionContext(this)
+            ctx.convertAndAppend(body)
+            return ctx.block
+        }
+}
+
+class StmtConversionContext(val methodCtx: MethodConversionContext) {
+    val statements: MutableList<Stmt> = mutableListOf()
+    val declarations: MutableList<Declaration> = mutableListOf()
+    val block = Seqn(statements, declarations)
+
+    fun convertAndAppend(stmt: FirStatement) {
+        stmt.accept(StmtConversionVisitor(), this)
+    }
+}
+
+class StmtConversionVisitor : FirVisitor<Exp?, StmtConversionContext>() {
+    override fun visitElement(element: FirElement, data: StmtConversionContext): Exp? {
+        throw Exception("StmtConversionVisitor should only be used to convert statements.")
+    }
+
+    override fun visitStatement(statement: FirStatement, data: StmtConversionContext): Exp? {
+        TODO("Not yet implemented for $statement (${statement.source.text})")
+    }
+
+    override fun visitReturnExpression(returnExpression: FirReturnExpression, data: StmtConversionContext): Exp? {
+        val expr = returnExpression.result.accept(this, data)
+        // TODO: respect return-based control flow
+        if (expr != null) {
+            val returnVar = data.methodCtx.returnVar ?: throw Exception("Expression returned in void function")
+            data.statements.add(Stmt.LocalVarAssign(returnVar.toLocalVar(), expr))
+        }
+        System.err.println("Visiting: $returnExpression")
+        return null
+    }
+
+    override fun visitBlock(block: FirBlock, data: StmtConversionContext): Exp? {
+        // TODO: allow blocks to return values
+        block.statements.forEach { it.accept(this, data) }
+        return null
+    }
+
+    override fun <T> visitConstExpression(constExpression: FirConstExpression<T>, data: StmtConversionContext): Exp? {
```

Comment:
Same thing for the `when` expression also here!

```kotlin
override fun <T> visitConstExpression(constExpression: FirConstExpression<T>, data: StmtConversionContext): Exp? 
 = when (constExpression.kind) {
    ConstantValueKind.Int -> Exp.IntLit((constExpression.value as Long).toInt().toScalaBigInt())
    else -> TODO("Not implemented yet")
}
```

Reply:
Done